engine/chunker/cached_chunker.go

```diff
@@ -0,0 +1,516 @@
+		cache.Clear()
+	})
+}
+
+// Close syncs the backing datastore but does not close it.
+// This is because cached entries chunker wraps an existing datastore and does
+// not construct it, and the wrapped datastore may be in use elsewhere.
+func (ls *CachedEntriesChunker) Close() error {
+	return ls.sync(context.TODO())
+}
+
+// restoreCache restores the cached entries from the backing datastore and cleans up the datastore
+// such that only chunks associated to the root of chains remain in the datastore.
+func (ls *CachedEntriesChunker) restoreCache(ctx context.Context) error {
+	// Query the root keys of entries chains.
+	q := dsq.Query{
+		Prefix:   rootKeyPrefix.String(),
+		KeysOnly: true,
+	}
+
+	results, err := ls.ds.Query(ctx, q)
+	if err != nil {
+		return err
+	}
+	defer results.Close()
+
+	// For each root key
+	var count int
+	for r := range results.Next() {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if r.Error != nil {
+			return fmt.Errorf("cannot read cache key: %w", r.Error)
+		}
+
+		// Extract the root link from its datastore key
+		rawKey := datastore.RawKey(r.Key)
+		l, err := ls.linkFromDsCachePrefixedKey(rawKey)
+		if err != nil {
+			return err
+		}
+
+		// List all of root's successive links by traversing the chain
+		links, err := ls.listEntriesChainLinks(ctx, l)
+		if err != nil {
+			return err
+		}
+
+		// Update in memory cache with root link and its list of links
+		err = ls.performOnCache(ctx, func(cache *lru.Cache) { cache.Add(l, links) })
+		if err != nil {
+			return err
+		}
+		count++
+	}
+
+	// If no root key is present in datastore, it means the cache should be empty
+	// Therefore, clear all keys in the datastore.
+	//
+	// This also makes sure that data cached using previous implementation of caching is cleared.
+	if count == 0 {
+		// Query all keys in datastore.
+		allKeys := dsq.Query{
+			KeysOnly: true,
+		}
+		allKeysResult, err := ls.ds.Query(ctx, allKeys)
+		if err != nil {
+			return err
+		}
+		defer allKeysResult.Close()
+
+		// For each key in datastore delete its corresponding value.
+		var prunedCount int
+		for r := range allKeysResult.Next() {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			if r.Error != nil {
+				return fmt.Errorf("cannot read cache key: %w", r.Error)
+			}
+			err := ls.ds.Delete(ctx, datastore.RawKey(r.Key))
+			if err != nil {
+				return err
+			}
+			prunedCount++
+		}
+
+		// If datastore was pruned, log informative message.
+		if prunedCount != 0 {
+			log.Infow("No caching metadata is persisted but datastore is non-empty; pruned lingering cache entries", "count", prunedCount)
+		}
+	} else {
+		log.Debugw("Cache restored successfully", "restoredCacheCount", ls.Len(), "capacity", ls.Cap())
+	}
+
+	// If the cache capacity was too small to restore all entries present, it means cache was
+	// evicted during restore and records were pruned as needed.
+	//
+	// Log an informative message to let the user know.
+	if ls.Cap() < count {
```

Comment:
Should this be moved into the above `else` block?  This does not seem useful if `count` is 0.